File: src/katas/javascript/kyu-6/kata-1.js
// DigitPower class with a single responsibility: to find the value of k for given n and p.
class DigitPower {
  constructor(n, p) {
    this.n = n;
    this.p = p;
  }

  // Method to find and return the value of k
  findK() {
    // Convert the number to a string and then to an array of digits
    const digits = Array.from(String(this.n), Number);

    // Calculate the sum of the digits taken to the successive powers of p
    let sum = 0;
    for (let i = 0; i < digits.length; i++) {
      sum += Math.pow(digits[i], this.p + i);
    }

    // Find k such that sum = n * k
    const k = sum / this.n;

    // If k is an integer, return k. Otherwise, return -1.
    return Number.isInteger(k) ? k : -1;
  }
}

// Testing the DigitPower class
const dig_pow = (n, p) => {
  const digitPower = new DigitPower(n, p);
  return digitPower.findK();
};

File: src/katas/python/kyu-6/kata-1.py
def dig_pow(n, p):
    # Step 1: Convert n to a list of its digits
    digits = [int(d) for d in str(n)]

    # Step 2: Calculate the sum of digits raised to the power p, p+1, ...
    total_sum = sum(d ** (p + i) for i, d in enumerate(digits))

    # Step 3: Check if total_sum is divisible by n
    k, remainder = divmod(total_sum, n)

    return k if remainder == 0 else -1
